### DISCUSSION TITLE
File collisions when storing the same instance twice

### DISCUSSION THREAD

**Discussion by or150** [dcm4che/dcm4che#1108] [Q&A]
**Describe the bug**
Hi,
I'm using dcm4che storescp in several high load windows servers.
In some scenarios the same SOP Instance is sent to the server from two or more sources at the same time.
This causes several random errors that happen on a daily basis:
* M-RENAME failures in log
* IOExceptions are returned to the sender
* Partial dicoms are being moved to the destination folder

**To Reproduce**
1. Run storescp on a windows server.
2. Send the same instance multiple times in parallel to the server (see script below).
4. The error will happen after

**Expected behavior**
No errors when sending the same instance twice.

**Version Information:**
 - OS: Windows
 - Version 5.24.1

**Additional context**
Python script I used to simulate the error:
```python
import os
import random
import subprocess
import time
from concurrent import futures
from concurrent.futures.thread import ThreadPoolExecutor

MAX_PARALLEL_STUDIES = 10
STUDIES_PER_HOUR = 3000.0*1000.0
SECONDS_IN_HOUR = 3600.0
STUDIES_PER_SECOND = STUDIES_PER_HOUR / SECONDS_IN_HOUR
TOLERANCE_SEC = 30

MAX_STUDIES_TO_SEND = 1000
STORESCP_TARGETS = ['store@127.0.0.1:4243']
DCM4CHE_PATH = os.path.expanduser(r'C:\dcm4che\bin')

pool_executor = ThreadPoolExecutor(max_workers=MAX_PARALLEL_STUDIES)


def send_study(study_path: str, storescp_target: str):
    subprocess.run([os.path.join(DCM4CHE_PATH, 'storescu'), "-c", storescp_target, study_path], check=False,
                   stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
    print(f'Sent: {study_path}')


def get_storescp_target():
    while True:
        for target in STORESCP_TARGETS:
            yield target


def get_studies_to_send():
    for _ in range(MAX_STUDIES_TO_SEND):
        yield "D:\Data\one_instance"


def main():
    start_time = time.time()
    end_time = time.time()

    studies_to_send = list(get_studies_to_send())
    random.shuffle(studies_to_send)
    sent_counter = 0
    sending = []
    for study_path, storescp_target in zip(studies_to_send, get_storescp_target()):
        future = pool_executor.submit(send_study, study_path, storescp_target)
        sending.append(future)
        if len(sending) >= MAX_PARALLEL_STUDIES:
            futures.wait(sending, return_when=futures.FIRST_COMPLETED)

        new_sending = []
        for f in sending:
            if f.done():
                sent_counter += 1
            else:
                new_sending.append(f)
        sending = new_sending

        end_time = time.time()
        passed_time_since_start = end_time - start_time
        expected_time_to_pass_since_start = sent_counter / STUDIES_PER_SECOND
        sleep_duration = expected_time_to_pass_since_start - passed_time_since_start - TOLERANCE_SEC
        if sleep_duration > 0:
            print(f'Sleeping for: {sleep_duration} seconds')
            time.sleep(sleep_duration)

    print(f'Took: {end_time - start_time} seconds')


if __name__ == '__main__':
    main()
```

**Comment by or150** [comment]
I solved the issue for my use cases with these changes: #1104

**Comment by gunterze** [comment]
Specify  `--filepath <pattern>` including `{rnd}` which will be replaced by random 8 hex digits (#384)

**Reply by or150** [reply]
Hi,

`--filepath` doesn't affect the temporary file:
https://github.com/dcm4che/dcm4che/blob/master/dcm4che-tool/dcm4che-tool-storescp/src/main/java/org/dcm4che3/tool/storescp/StoreSCP.java#L106

Which is the cause for:
Corrupted instance files (instances with half the data).
Some of the errors that the storescu's are receiving when the collision happens at very exact timings (the rest are what I added the retries for).

It also doesn't solve my issue. I don't want to store the file twice. This is a lot of redundant storage. I don't mind it being overwritten, but the storage space is scarce in my environments.

My change isn't a breaking one, can you consider it again? I don't want to keep maintaining a fork because of these bugs.

Thank you

**Reply by gunterze** [reply]
Fixed by d2d43d4

**Reply by or150** [reply]
And about the 2nd issue? Where the rename fails? Do you have a suggestion how to solve it?

Also, part of the change did improve the error we get when rename fails:
```java
Files.move(from.toPath(), dest.toPath(), StandardCopyOption.REPLACE_EXISTING)
```
Instead of:
```java
        if (!dest.getParentFile().mkdirs())
            dest.delete();
        if (!from.renameTo(dest))
            throw new IOException("Failed to rename " + from + " to " + dest);
```
Raises an indicative error with the cause instead of just "failed to rename".

**Reply by gunterze** [reply]
pushed ce6ce66fa

**Reply by fmarot** [reply]
Hello, I investigated today and realized I was having the same error. It seems like my "partner pacs" sends me the same file multiple times simultaneously. I solved it by putting a "synchronized" before the "renameTo" method. It is a quick and dirty solution, maybe suboptimal performance-wise, but at least I do not get errors anymore.
I think it would be wise to have it "synchronized" by default in DCM4CHE because errors can cause headaches for users.

Do you want a merge request ?

**Reply by fmarot** [reply]
Hello again, I opend an issue and proposed a fix as a MR: https://github.com/dcm4che/dcm4che/issues/1189

**Reply by or150** [reply]
Hi,

My suggested fix included a second part, with retries on the rename.

The reasons I went with retries instead of locks were:
- It doesn't affect performance as much as locks
- It allows running two listeners that write to the same directory.

I did performance tests on a very high velocity throughput that represents one of our use cases and was able to reduce the issue to a practically nonexistent level with 3 retries.

@gunterze If you'd reconsider merging that part too, It'll be great.
